application/src/main/java/org/thingsboard/server/service/cf/ctx/state/GeofencingCalculatedFieldState.java

```
@@ -0,0 +1,207 @@
+                            throw new IllegalArgumentException("Unsupported argument entry type for " + key + " argument: " + newEntry.getType() + ". " +
+                                                               "Only GEOFENCING type is allowed.");
+                        }
+                        arguments.put(key, geofencingArgumentEntry);
+                        yield true;
+                    }
+                };
+            } else {
+                entryUpdated = existingEntry.updateEntry(newEntry);
+            }
+            if (entryUpdated) {
+                stateUpdated = true;
+            }
+        }
+        return stateUpdated;
+    }
+
+    @Override
+    public ListenableFuture<CalculatedFieldResult> performCalculation(EntityId entityId, CalculatedFieldCtx ctx) {
+        double latitude = (double) arguments.get(ENTITY_ID_LATITUDE_ARGUMENT_KEY).getValue();
+        double longitude = (double) arguments.get(ENTITY_ID_LONGITUDE_ARGUMENT_KEY).getValue();
+        Coordinates entityCoordinates = new Coordinates(latitude, longitude);
+
+        var geofencingCfg = (GeofencingCalculatedFieldConfiguration) ctx.getCalculatedField().getConfiguration();
+        Map<String, ZoneGroupConfiguration> zoneGroups = geofencingCfg
+                .getZoneGroups()
+                .stream()
+                .collect(Collectors.toMap(ZoneGroupConfiguration::getName, Function.identity()));
+
+        ObjectNode resultNode = JacksonUtil.newObjectNode();
+        List<ListenableFuture<Boolean>> relationFutures = new ArrayList<>();
+
+        getGeofencingArguments().forEach((argumentKey, argumentEntry) -> {
+            ZoneGroupConfiguration zoneGroupCfg = zoneGroups.get(argumentKey);
+            if (zoneGroupCfg == null) {
+                log.error("[{}][{}] Zone group config is missing for the {}", entityId, ctx.getCalculatedField().getId(), argumentKey);
+                return;
```

Also, why ignore the error?